packages/bun-uws/src/HttpResponse.h

```diff
@@ -43,561 +43,624 @@ static const char *HTTP_200_OK = "200 OK";
+    Super::write("\r\n", 2);
+    return this;
+  }
+
+  /* Write an HTTP header with unsigned int value */
+  HttpResponse *writeHeader(std::string_view key, uint64_t value) {
+    writeStatus(HTTP_200_OK);
+
+    Super::write(key.data(), (int)key.length());
+    Super::write(": ", 2);
+    writeUnsigned64(value);
+    Super::write("\r\n", 2);
+    return this;
+  }
+
+  /* End without a body (no content-length) or end with a spoofed
+   * content-length. */
+  void
+  endWithoutBody(std::optional<size_t> reportedContentLength = std::nullopt,
+                 bool closeConnection = false) {
+    if (reportedContentLength.has_value()) {
+      internalEnd({nullptr, 0}, reportedContentLength.value(), false, true,
+                  closeConnection);
+    } else {
+      internalEnd({nullptr, 0}, 0, false, false, closeConnection);
+    }
+  }
+
+  /* End the response with an optional data chunk. Always starts a timeout. */
+  void end(std::string_view data = {}, bool closeConnection = false) {
+    internalEnd(data, data.length(), false, true, closeConnection);
+  }
+
+  /* Try and end the response. Returns [true, true] on success.
+   * Starts a timeout in some cases. Returns [ok, hasResponded] */
+  std::pair<bool, bool> tryEnd(std::string_view data, uintmax_t totalSize = 0,
+                               bool closeConnection = false) {
+    return {internalEnd(data, totalSize, true, true, closeConnection),
+            hasResponded()};
+  }
+
+  /* Write the end of chunked encoded stream */
+  bool sendTerminatingChunk(bool closeConnection = false) {
+    writeStatus(HTTP_200_OK);
+    HttpResponseData<SSL> *httpResponseData = getHttpResponseData();
+    if (!(httpResponseData->state & HttpResponseData<SSL>::HTTP_WRITE_CALLED)) {
+      /* Write mark on first call to write */
+      writeMark();
+
+      writeHeader("Transfer-Encoding", "chunked");
+      httpResponseData->state |= HttpResponseData<SSL>::HTTP_WRITE_CALLED;
     }
 
-    /* Write parts of the response in chunking fashion. Starts timeout if failed. */
-    bool write(std::string_view data) {
-        writeStatus(HTTP_200_OK);
-
-        /* Do not allow sending 0 chunks, they mark end of response */
-        if (!data.length()) {
-            /* If you called us, then according to you it was fine to call us so it's fine to still call us */
-            return true;
-        }
-
-        HttpResponseData<SSL> *httpResponseData = getHttpResponseData();
-
-        if (!(httpResponseData->state & HttpResponseData<SSL>::HTTP_WRITE_CALLED)) {
-            /* Write mark on first call to write */
-            writeMark();
-
-            writeHeader("Transfer-Encoding", "chunked");
-            httpResponseData->state |= HttpResponseData<SSL>::HTTP_WRITE_CALLED;
-        }
+    /*  this will be sent always when state is HTTP_WRITE_CALLED inside internalEnd, so no need to write the terminating 0 chunk here
+     Super::write("\r\n0\r\n\r\n", 7); */
```

Comment:
the only change is that we comment on this part because we guarantee that state is always HTTP_WRITE_CALLED